Byte extract lowering for vectors and array cleanup [blocks: #2068, #4187]

tautschnig opened the pull request:
Only the last three commits are new, the rest is borrowed from #4123.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
byte_extract lowering of arrays: optimise for constant index

No need to construct index_exprt when we can just pick the operand out of a
vector.

tautschnig committed:
byte_extract lowering of vectors and array cleanup

Refactor the code used for arrays to make it re-usable for vectors and arrays.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c21e390f3bced900d4cc8e433cdc029e303e191b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100140199

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -499,8 +499,21 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    simplify(offset_i, ns);
+
+    mp_integer index = 0;
+    if(
```
Comment that this amounts to "if `offset_i` is constant and in range". Also perhaps invert the if/else? I'd usually expect `if(can_do_good_thing) good_thing; else fallback;`

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+
+/// rewrite an array or vector into its individual bytes
+/// \param src: array/vector to unpack
+/// \param src_size: array/vector size
```
Clarify this could be unknown, and we need either this or max_bytes?

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+    if(!num_elements)
+    {
+      if(*max_bytes_int % el_bytes != 0)
+        *max_bytes_int += el_bytes;
```
Use a differently-named variable when we change units (`max_size` to match `src_size` I guess)

tautschnig replied:
Done (I added a new `optionalt<mp_integer> max_elements;`)

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+    // turn bytes into elements
+    if(!num_elements)
+    {
+      if(*max_bytes_int % el_bytes != 0)
```
Consider a util for divide-rounding-up like this

tautschnig replied:
Now done in a single instruction.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+  // array/vector is unknown; if the element_width was usable above this will
+  // have been turned into a number of elements already
+  if(!num_elements)
+    num_elements = max_bytes_int;
```
Eep! Definitely rename `max_bytes_int`

tautschnig replied:
Yep, done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+      element = src_simp.operands()[index_int];
+    }
+
+    exprt sub = unpack_rec(element, little_endian, max_bytes, ns, true);
```
This is surprising, what's going on here with an element expanding into a multi-element array-or-vector and the result being flattened?

tautschnig replied:
I've added a comment: the task of `unpack_rec` is to produce an array of bytes, thus we need to turn each of the original array elements into bytes as well.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+/// \param max_bytes: if not nil, use as upper bound of the number of bytes to
+///                   unpack
+/// \param ns: namespace for type lookups
+/// \param unpack_byte_array: if true, return unmodified src iff it is a an
```
typo `a an`

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+  {
+    exprt element;
+
+    if(src_simp.id() != ID_array || i >= src_simp.operands().size())
```
Flip the `if/else` so the special case comes first?

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+///                   unpack
+/// \param ns: namespace for type lookups
+/// \param [out] dest_array: target array_exprt to hold unpacked elements
+static void unpack_array_vector(
```
AFAICT `dest_array` is only non-empty on recursion, in which case split this into a `_rec` with an out-append parameter and a root function that returns an `array_exprt`

tautschnig replied:
Done, this makes things a lot clearer I think.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,95 @@ Author: Daniel Kroening, kroening@kroening.com
+  {
+    exprt element;
+
+    if(src_simp.id() != ID_array || i >= src_simp.operands().size())
```
Can't the good case be done for `ID_vector` too?

tautschnig replied:
Yep, fixed just now while addressing your earlier comment :-)

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -30,6 +31,7 @@ static exprt unpack_rec(
 /// \param src_size: array/vector size
 /// \param element_width: bit width of array/vector elements
 /// \param little_endian: true, iff assumed endianness is little-endian
+/// \param offset: if not nil, use as lower bound for non-nil bytes to return
```
Odd phrasing! How about
`if not nil, bytes prior to this offset will be filled with nil values`

tautschnig replied:
Thank you :-)

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -39,6 +41,7 @@ static void unpack_array_vector(
   const exprt &src_size,
   const mp_integer &element_width,
   bool little_endian,
+  const exprt &offset,
```
rename `offset_bytes`

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -67,6 +70,17 @@ static void unpack_array_vector(
       *max_bytes_int /= el_bytes;
     }
+
+    if(auto offset_bytes = numeric_cast<mp_integer>(offset))
```
rename `offset_bytes_int`

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -102,6 +117,7 @@ static void unpack_array_vector(
 /// rewrite an object into its individual bytes
 /// \param src: object to unpack
 /// \param little_endian: true, iff assumed endianness is little-endian
+/// \param offset: if not nil, use as lower bound for non-nil bytes to return
```
Phrasing as above

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -182,11 +203,38 @@ static exprt unpack_rec(
         throw non_byte_alignedt(struct_type, comp, *element_width);
       }
 
+      exprt new_offset = nil_exprt();
```
`s/new/element`?

tautschnig replied:
Renamed to `offset_in_member`

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -182,11 +203,38 @@ static exprt unpack_rec(
+      {
+        *new_offset_int -= member_offset_bits / 8;
+        if(*new_offset_int >= 0)
+          new_offset = from_integer(*new_offset_int, offset.type());
```
Comment that the `else` is intentionally missing

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -182,11 +203,38 @@ static exprt unpack_rec(
+      if(src.id() == ID_struct)
+        simplify(member, ns);
+
+      exprt sub =
```
`sub` defined but not used?

tautschnig replied:
Its operands are used.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -294,9 +342,12 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
         ns);
 
+  exprt lb = src.offset();
+  if(!lb.is_constant())
```
This constraint suggests passing `optional<constant_exprt>` rather than `exprt`

tautschnig replied:
Fair enough - I'll add an extra commit that refactors this.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -246,7 +246,7 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
         plus_exprt new_offset(
           unpacked.offset(),
-          from_integer(i * (*element_width), unpacked.offset().type()));
+          from_integer(i * (*element_width) / 8, unpacked.offset().type()));
```
Consider using `_bytes` and _bits` suffixes to discourage this sort of problem in future

tautschnig replied:
Will do in an additional commit.

tautschnig replied:
Done.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f62cfae3f306f3d9640cc45c5ee2727aaf00d0de).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100907943

tautschnig committed:
byte_extract lowering: also use a lower bound

We already limited extraction at an uppper bound, but when we know the offset we
can also start as late as possible. For structs, for example, we now do not
expand components that will be irrelevant to the final result.

tautschnig committed:
byte-operator lowering: rename element_width to {element,component}_bits

This makes the unit of measurement explicit and thus should help avoid
bits/bytes confusion.

tautschnig committed:
byte_extract lowering: Use optionalt<mp_integer> instead of nil_exprt

This makes the interface much more explicit about what is or isn't supported at
moment.

tautschnig commented:
@smowton All comments addressed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -18,13 +18,102 @@ Author: Daniel Kroening, kroening@kroening.com
+    // turn bytes into elements
+    if(!num_elements)
+    {
+      // round up to nearest multiple of el_bytes
```
7 rounded up to the nearest multiple of 4 is 8, not 2. How about `// max_bytes_int / el_bytes, rounding up`

tautschnig replied:
Comment adjusted.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 88365231f5ea9e654dcaa60ed981b1b6f9a3d70a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100944534

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -111,7 +111,7 @@ static array_exprt unpack_array_vector(
     // array of bytes left
     exprt sub =
-      unpack_rec(element, little_endian, nil_exprt(), max_bytes, ns, true);
+      unpack_rec(element, little_endian, nullopt, max_bytes, ns, true);
```
Usually `{}` rather than `nullopt`

tautschnig replied:
We seem to have a small number of uses, but agreed that `{}` is more common and thus to be used.

smowton reviewed (approved):
Looks good, thanks! One comment I think is wrong and a tiny nitpick -- one other thing: can we avoid the back-and-forth between `constant_exprt` and `optional<mp_integer>` that happens when we recurse into a member? That seems a little error-prone, as well as expensive.

tautschnig commented:
> can we avoid the back-and-forth between `constant_exprt` and `optional<mp_integer>` that happens when we recurse into a member? That seems a little error-prone, as well as expensive.

Using `optionalt<mp_integer>` makes things again much simpler :-) - thanks!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bd307906f68c1da07ec138999e871b187092c7ec).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100967585
